### PR TITLE
addmod, mulmod, conditionals, parentheses, testharness, bug fixes

### LINKED ISSUE #20: Solidity feature: Conditionals
E.g. `x ? y : z`

### PR DESCRIPTION
**Pull request by dwightguth**
Fixes #20 

Contains:
  * support for addmod and mulmod instructions
  * support for conditional expressions
  * support for TupleExpressions with one component (ie, parentheses)
  * fixes some bugs in the compiler in preparation for running end to end tests
  * add an assertion to fail the test on local variables of function type
  * improved test harness

### CODE REVIEW

File: libsolidity/codegen/IeleCompiler.cpp [hunk 1]
```diff
@@ -380,7 +386,51 @@ bool IeleCompiler::visit(const InlineAssembly &inlineAssembly) {
...
+  iele::IeleValue *ConditionValue = compileExpression(condition.condition());
+  assert(ConditionValue && "IeleCompiler: failed to compile conditional condition.");
+
+  // The condition target block is the if-true block.
+  iele::IeleBlock *CondTargetBlock =
+    iele::IeleBlock::Create(&Context, "if.true");
+
+  // Connect the condition block with a conditional jump to the condition target
+  // block.
+  connectWithConditionalJump(ConditionValue, ConditionBlock, CondTargetBlock);
```
Comment (theo25) [inline libsolidity/codegen/IeleCompiler.cpp]: Here you should change `ConditionBlock` to `CompilingBlock` for the case where the previously called `compileExpression(condition.condition())` contained another conditional and as a result has changed the current compiling block. See the final fixed implementation for the if statement (tips of the if-fix branch).

Comment (theo25) [inline libsolidity/codegen/IeleCompiler.cpp]: Also I think, the way you have implemented it (e.g jump to if-true block or fall through for if-false) you need to invert the condition.

Comment (dwightguth) [reply libsolidity/codegen/IeleCompiler.cpp]: no? that's not what IfStatement does. It does that only if there is no false branch of the if statement, but conditional expressions always have both a true and false branch.

Also the end to end tests pass right now, which they certainly wouldn't if this were backwards.

Comment (theo25) [reply libsolidity/codegen/IeleCompiler.cpp]: sorry my bad, you are right

**Pull request merged**